// @torihada-inc/fanme-sdk-backend/dist/appApi.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SellerAppItemPrice = exports.SellerAppItemId = exports.SellerAppItem = exports.SellerApp = exports.BaseAppItem = exports.BaseApp = exports.notNullish = void 0;
const typescript_json_serializer_1 = require("typescript-json-serializer");
function notNullish(item) {
    return !(0, typescript_json_serializer_1.isNullish)(item);
}
exports.notNullish = notNullish;
class AppApi {
    constructor(axios, baseAppId, sellerUserId, sellerAppId) {
        this.multipartConfig = { headers: { 'content-type': 'multipart/form-data' } };
        this.serializer = new typescript_json_serializer_1.JsonSerializer();
        this.axios = axios;
        this.baseAppId = baseAppId;
        this.sellerUserId = sellerUserId;
        this.sellerAppId = sellerAppId;
    }
    // /*
    //  * アプリケーション取得
    //  * 戻り値: 開発者の登録したアプリケーション
    //  */
    // public async getApps(): Promise<Array<any>> {
    //     return await this.axios.get(`/sdk/${this.sellerUserId}/available`)
    //         .then(res => {
    //             const r = this.serializer.deserializeObjectArray(res.data, GetAppResponse)
    //             return r?.filter(notNullish) ?? Promise.reject()
    //         })
    //         .catch(err => {
    //             return Promise.reject(err)
    //         })
    // }
    /*
     * アプリケーション取得
     * 戻り値: 開発者の登録したアプリケーション
     */
    getBaseApp() {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.axios.get(`/sdk/${this.sellerUserId}/available/${this.baseAppId}`)
                .then(res => {
                const r = this.serializer.deserializeObject(res.data, BaseApp);
                return r !== null && r !== void 0 ? r : Promise.reject();
            })
                .catch(err => {
                return Promise.reject(err);
            });
        });
    }
    /*
     * アプリケーション商品取得
     * 戻り値: 開発者の登録したアプリケーション商品
     */
    getBaseAppItems() {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.axios.get(`/sdk/${this.sellerUserId}/available/${this.baseAppId}/items`)
                .then(res => {
                var _a;
                const r = this.serializer.deserializeObjectArray(res.data, BaseAppItem);
                return (_a = r === null || r === void 0 ? void 0 : r.filter(notNullish)) !== null && _a !== void 0 ? _a : Promise.reject();
            })
                .catch(err => {
                return Promise.reject(err);
            });
        });
    }
    // /*
    //  * アプリケーション取得
    //  * 戻り値: 自分の登録したアプリケーション
    //  */
    // public async getMyApps(): Promise<Array<GetSellerAppResponse>> {
    //     return await this.axios.get(`/sdk/${this.sellerUserId}/mine`)
    //         .then(res => {
    //             const r = this.serializer.deserializeObjectArray(res.data, GetSellerAppResponse)
    //             return r?.filter(notNullish) ?? Promise.reject()
    //         })
    //         .catch(err => {
    //             return Promise.reject(err)
    //         })
    // }
    /*
     * アプリケーション取得
     * 戻り値: 自分の登録したアプリケーション
     */
    getSellerApp() {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.axios.get(`/sdk/${this.sellerUserId}/mine/${this.sellerAppId}`)
                .then(res => {
                const r = this.serializer.deserializeObject(res.data, SellerApp);
                return r !== null && r !== void 0 ? r : Promise.reject();
            })
                .catch(err => {
                return Promise.reject(err);
            });
        });
    }
    /*
     * アプリケーション商品取得
     * 戻り値: 自分の登録したアプリケーション
     */
    getSellerAppItems() {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.axios.get(`/sdk/${this.sellerUserId}/${this.sellerAppId}/items`)
                .then(res => {
                var _a;
                const r = this.serializer.deserializeObjectArray(res.data, SellerAppItem);
                return (_a = r === null || r === void 0 ? void 0 : r.filter(notNullish)) !== null && _a !== void 0 ? _a : Promise.reject();
            })
                .catch(err => {
                return Promise.reject(err);
            });
        });
    }
    /*
     * アプリケーション商品取得
     * 戻り値: 自分の登録したアプリケーション
     */
    getSellerAppItem(sellerAppItemId) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.axios.get(`/sdk/${this.sellerUserId}/${this.sellerAppId}/items/${sellerAppItemId}`)
                .then(res => {
                const r = this.serializer.deserializeObject(res.data, SellerAppItem);
                return r !== null && r !== void 0 ? r : Promise.reject();
            })
                .catch(err => {
                return Promise.reject(err);
            });
        });
    }
    /*
     * アプリケーション商品新規作成
     * 戻り値: 作成したsellerAppId
     */
    createSellerAppItem(baseAppItemId, name, sortOrder, shortName, description, iconUrl, imageUrl1, imageUrl2, imageUrl3, imageUrl4, imageUrl5, available = true) {
        return __awaiter(this, void 0, void 0, function* () {
            let Request = class Request {
                constructor(baseAppItemId, sellerAppId, name, description, iconUrl, imageUrl1, imageUrl2, imageUrl3, imageUrl4, imageUrl5, sortOrder, available) {
                    this.baseAppItemId = baseAppItemId;
                    this.sellerAppId = sellerAppId;
                    this.name = name;
                    this.description = description;
                    this.iconUrl = iconUrl;
                    this.imageUrl1 = imageUrl1;
                    this.imageUrl2 = imageUrl2;
                    this.imageUrl3 = imageUrl3;
                    this.imageUrl4 = imageUrl4;
                    this.imageUrl5 = imageUrl5;
                    this.sortOrder = sortOrder;
                    this.available = available;
                }
            };
            Request = __decorate([
                (0, typescript_json_serializer_1.JsonObject)(),
                __param(0, (0, typescript_json_serializer_1.JsonProperty)({ name: "app_item_id", required: true })),
                __param(1, (0, typescript_json_serializer_1.JsonProperty)({ name: "seller_app_id", required: true })),
                __param(2, (0, typescript_json_serializer_1.JsonProperty)({ name: "name", required: true })),
                __param(3, (0, typescript_json_serializer_1.JsonProperty)({ name: "description" })),
                __param(4, (0, typescript_json_serializer_1.JsonProperty)({ name: "icon_url" })),
                __param(5, (0, typescript_json_serializer_1.JsonProperty)({ name: "image_url1" })),
                __param(6, (0, typescript_json_serializer_1.JsonProperty)({ name: "image_url2" })),
                __param(7, (0, typescript_json_serializer_1.JsonProperty)({ name: "image_url3" })),
                __param(8, (0, typescript_json_serializer_1.JsonProperty)({ name: "image_url4" })),
                __param(9, (0, typescript_json_serializer_1.JsonProperty)({ name: "image_url5" })),
                __param(10, (0, typescript_json_serializer_1.JsonProperty)({ name: "sort_order", required: true })),
                __param(11, (0, typescript_json_serializer_1.JsonProperty)({ name: "available", required: true })),
                __metadata("design:paramtypes", [Number, Number, String, String, String, String, String, String, String, String, Number, Boolean])
            ], Request);
            const request = new Request(baseAppItemId, this.sellerAppId, name, description, iconUrl, imageUrl1, imageUrl2, imageUrl3, imageUrl4, imageUrl5, sortOrder, available);
            return yield this.axios.post(`/sdk/${this.sellerUserId}/${this.sellerAppId}/items/register`, this.serializer.serialize(request))
                .then(res => {
                const r = this.serializer.deserializeObject(res.data, SellerAppItemId);
                return r !== null && r !== void 0 ? r : Promise.reject();
            })
                .catch(err => {
                return Promise.reject(err);
            });
        });
    }
    /*
     * アプリケーション商品更新
     * 戻り値: なし
     */
    updateSellerAppItem(sellerAppItemId, name, sortOrder, shortName, description, iconUrl, imageUrl1, imageUrl2, imageUrl3, imageUrl4, imageUrl5, available) {
        return __awaiter(this, void 0, void 0, function* () {
            let Request = class Request {
                constructor(sellerAppItemId, name, description, iconUrl, imageUrl1, imageUrl2, imageUrl3, imageUrl4, imageUrl5, sortOrder, available) {
                    this.sellerAppItemId = sellerAppItemId;
                    this.name = name;
                    this.description = description;
                    this.iconUrl = iconUrl;
                    this.imageUrl1 = imageUrl1;
                    this.imageUrl2 = imageUrl2;
                    this.imageUrl3 = imageUrl3;
                    this.imageUrl4 = imageUrl4;
                    this.imageUrl5 = imageUrl5;
                    this.sortOrder = sortOrder;
                    this.available = available;
                }
            };
            Request = __decorate([
                (0, typescript_json_serializer_1.JsonObject)(),
                __param(0, (0, typescript_json_serializer_1.JsonProperty)({ name: "seller_app_item_id", required: true })),
                __param(1, (0, typescript_json_serializer_1.JsonProperty)({ name: "name" })),
                __param(2, (0, typescript_json_serializer_1.JsonProperty)({ name: "description" })),
                __param(3, (0, typescript_json_serializer_1.JsonProperty)({ name: "icon_url" })),
                __param(4, (0, typescript_json_serializer_1.JsonProperty)({ name: "image_url1" })),
                __param(5, (0, typescript_json_serializer_1.JsonProperty)({ name: "image_url2" })),
                __param(6, (0, typescript_json_serializer_1.JsonProperty)({ name: "image_url3" })),
                __param(7, (0, typescript_json_serializer_1.JsonProperty)({ name: "image_url4" })),
                __param(8, (0, typescript_json_serializer_1.JsonProperty)({ name: "image_url5" })),
                __param(9, (0, typescript_json_serializer_1.JsonProperty)({ name: "sort_order" })),
                __param(10, (0, typescript_json_serializer_1.JsonProperty)({ name: "available" })),
                __metadata("design:paramtypes", [Number, String, String, String, String, String, String, String, String, Number, Boolean])
            ], Request);
            const request = new Request(sellerAppItemId, name, description, iconUrl, imageUrl1, imageUrl2, imageUrl3, imageUrl4, imageUrl5, sortOrder, available);
            return yield this.axios.post(`/sdk/${this.sellerUserId}/${this.sellerAppId}/items/update`, this.serializer.serialize(request))
                .then(res => {
                const r = this.serializer.deserializeObject(res.data, SellerAppItemId);
                return r !== null && r !== void 0 ? r : Promise.reject();
            })
                .catch(err => {
                return Promise.reject(err);
            });
        });
    }
    /*
     * アプリケーション商品削除
     * 戻り値: なし
     */
    removeSellerAppItem(sellerAppItemId) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.axios.post(`/sdk/${this.sellerUserId}/${this.sellerAppId}/items/delete`, this.serializer.serialize({
                "seller_app_item_id": sellerAppItemId,
            }))
                .then(res => {
                const r = this.serializer.deserializeObject(res.data, SellerAppItemId);
                return r !== null && r !== void 0 ? r : Promise.reject();
            })
                .catch(err => {
                return Promise.reject(err);
            });
        });
    }
    /*
     * アプリケーション商品価格取得
     * 戻り値: 商品価格一覧
     */
    getSellerAppItemPrices(appItemId) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.axios.get(`/sdk/${this.sellerUserId}/${this.sellerAppId}/items/${appItemId}/prices`)
                .then(res => {
                var _a;
                const r = this.serializer.deserializeObjectArray(res.data, SellerAppItemPrice);
                return (_a = r === null || r === void 0 ? void 0 : r.filter(notNullish)) !== null && _a !== void 0 ? _a : Promise.reject();
            })
                .catch(err => {
                return Promise.reject(err);
            });
        });
    }
    /*
     * アプリケーション商品新規作成
     * 戻り値: なし
     */
    createSellerAppItemPrices(sellerAppItemId, prices) {
        return __awaiter(this, void 0, void 0, function* () {
            let Request = class Request {
                constructor(sellerAppItemId, prices) {
                    this.sellerAppItemId = sellerAppItemId;
                    this.prices = prices;
                }
            };
            Request = __decorate([
                (0, typescript_json_serializer_1.JsonObject)(),
                __param(0, (0, typescript_json_serializer_1.JsonProperty)({ name: "seller_app_item_id", required: true })),
                __param(1, (0, typescript_json_serializer_1.JsonProperty)({ name: "prices", required: true, type: (Array) })),
                __metadata("design:paramtypes", [Number, Array])
            ], Request);
            let RequestPrice = class RequestPrice {
                constructor(price, name, description) {
                    this.price = price;
                    this.name = name;
                    this.description = description;
                }
            };
            RequestPrice = __decorate([
                (0, typescript_json_serializer_1.JsonObject)(),
                __param(0, (0, typescript_json_serializer_1.JsonProperty)({ name: "price", required: true })),
                __param(1, (0, typescript_json_serializer_1.JsonProperty)({ name: "name" })),
                __param(2, (0, typescript_json_serializer_1.JsonProperty)({ name: "description" })),
                __metadata("design:paramtypes", [Number, String, String])
            ], RequestPrice);
            const request = new Request(sellerAppItemId, prices.map((value) => {
                return new RequestPrice(value.price, value.name, value.description);
            }));
            return yield this.axios.post(`/sdk/${this.sellerUserId}/${this.sellerAppId}/items/${sellerAppItemId}/prices/register`, this.serializer.serialize(request))
                .then(res => {
                return true;
            })
                .catch(err => {
                return Promise.reject(err);
            });
        });
    }
    /*
     * アプリケーション商品更新
     * 戻り値: なし
     */
    updateSellerAppItemPrices(appItemId, prices) {
        return __awaiter(this, void 0, void 0, function* () {
            let Request = class Request {
                constructor(sellerAppItemId, prices) {
                    this.sellerAppItemId = sellerAppItemId;
                    this.prices = prices;
                }
            };
            Request = __decorate([
                (0, typescript_json_serializer_1.JsonObject)(),
                __param(0, (0, typescript_json_serializer_1.JsonProperty)({ name: "seller_app_item_id", required: true })),
                __param(1, (0, typescript_json_serializer_1.JsonProperty)({ name: "prices", required: true, type: (Array) })),
                __metadata("design:paramtypes", [Number, Array])
            ], Request);
            let RequestPrice = class RequestPrice {
                constructor(sellerAppItemPriceId, price, name, description) {
                    this.sellerAppItemPriceId = sellerAppItemPriceId;
                    this.price = price;
                    this.name = name;
                    this.description = description;
                }
            };
            RequestPrice = __decorate([
                (0, typescript_json_serializer_1.JsonObject)(),
                __param(0, (0, typescript_json_serializer_1.JsonProperty)({ name: "seller_app_item_price_id", required: true })),
                __param(1, (0, typescript_json_serializer_1.JsonProperty)({ name: "price", required: true })),
                __param(2, (0, typescript_json_serializer_1.JsonProperty)({ name: "name" })),
                __param(3, (0, typescript_json_serializer_1.JsonProperty)({ name: "description" })),
                __metadata("design:paramtypes", [Number, Number, String, String])
            ], RequestPrice);
            const pricesReq = new Request(appItemId, prices.map((value) => {
                return new RequestPrice(value.sellerAppItemPriceId, value.price, value.name, value.description);
            }));
            return yield this.axios.post(`/sdk/${this.sellerUserId}/${this.sellerAppId}/items/${appItemId}/prices/update`, this.serializer.serialize(pricesReq))
                .then(res => {
                return true;
            })
                .catch(err => {
                return Promise.reject(err);
            });
        });
    }
}
exports.default = AppApi;
let BaseApp = class BaseApp {
    constructor(baseAppId, name, description, detailDescription, iconUrl, notice, imageUrl1, imageUrl2, imageUrl3, imageUrl4, imageUrl5, siteUrl, isDeveloping) {
        this.baseAppId = baseAppId;
        this.name = name;
        this.description = description;
        this.detailDescription = detailDescription;
        this.iconUrl = iconUrl;
        this.notice = notice;
        this.imageUrl1 = imageUrl1;
        this.imageUrl2 = imageUrl2;
        this.imageUrl3 = imageUrl3;
        this.imageUrl4 = imageUrl4;
        this.imageUrl5 = imageUrl5;
        this.siteUrl = siteUrl;
        this.isDeveloping = isDeveloping;
    }
};
BaseApp = __decorate([
    (0, typescript_json_serializer_1.JsonObject)(),
    __param(0, (0, typescript_json_serializer_1.JsonProperty)({ name: "app_id", required: true })),
    __param(1, (0, typescript_json_serializer_1.JsonProperty)({ name: "name", required: true })),
    __param(2, (0, typescript_json_serializer_1.JsonProperty)({ name: "description" })),
    __param(3, (0, typescript_json_serializer_1.JsonProperty)({ name: "detail_description" })),
    __param(4, (0, typescript_json_serializer_1.JsonProperty)({ name: "icon_url" })),
    __param(5, (0, typescript_json_serializer_1.JsonProperty)({ name: "notice" })),
    __param(6, (0, typescript_json_serializer_1.JsonProperty)({ name: "image_url1" })),
    __param(7, (0, typescript_json_serializer_1.JsonProperty)({ name: "image_url2" })),
    __param(8, (0, typescript_json_serializer_1.JsonProperty)({ name: "image_url3" })),
    __param(9, (0, typescript_json_serializer_1.JsonProperty)({ name: "image_url4" })),
    __param(10, (0, typescript_json_serializer_1.JsonProperty)({ name: "image_url5" })),
    __param(11, (0, typescript_json_serializer_1.JsonProperty)({ name: "site_url" })),
    __param(12, (0, typescript_json_serializer_1.JsonProperty)({ name: "is_developing" })),
    __metadata("design:paramtypes", [Number, String, String, String, String, String, String, String, String, String, String, String, Boolean])
], BaseApp);
exports.BaseApp = BaseApp;
let BaseAppItem = class BaseAppItem {
    constructor(baseAppItemId, baseAppId, name, description, detailDescription, iconUrl, notice, imageUrl1, imageUrl2, imageUrl3, imageUrl4, imageUrl5, appMarginRate, tenantMarginRate, minPrice, maxPrice, priceType, paymentType, tags, digitalContent) {
        this.baseAppItemId = baseAppItemId;
        this.baseAppId = baseAppId;
        this.name = name;
        this.description = description;
        this.detailDescription = detailDescription;
        this.iconUrl = iconUrl;
        this.notice = notice;
        this.imageUrl1 = imageUrl1;
        this.imageUrl2 = imageUrl2;
        this.imageUrl3 = imageUrl3;
        this.imageUrl4 = imageUrl4;
        this.imageUrl5 = imageUrl5;
        this.appMarginRate = appMarginRate;
        this.tenantMarginRate = tenantMarginRate;
        this.minPrice = minPrice;
        this.maxPrice = maxPrice;
        this.priceType = priceType;
        this.paymentType = paymentType;
        this.tags = tags;
        this.digitalContent = digitalContent;
    }
};
BaseAppItem = __decorate([
    (0, typescript_json_serializer_1.JsonObject)(),
    __param(0, (0, typescript_json_serializer_1.JsonProperty)({ name: "app_item_id", required: true })),
    __param(1, (0, typescript_json_serializer_1.JsonProperty)({ name: "app_id", required: true })),
    __param(2, (0, typescript_json_serializer_1.JsonProperty)({ name: "name", required: true })),
    __param(3, (0, typescript_json_serializer_1.JsonProperty)({ name: "description" })),
    __param(4, (0, typescript_json_serializer_1.JsonProperty)({ name: "detail_description" })),
    __param(5, (0, typescript_json_serializer_1.JsonProperty)({ name: "icon_url" })),
    __param(6, (0, typescript_json_serializer_1.JsonProperty)({ name: "notice" })),
    __param(7, (0, typescript_json_serializer_1.JsonProperty)({ name: "image_url1" })),
    __param(8, (0, typescript_json_serializer_1.JsonProperty)({ name: "image_url2" })),
    __param(9, (0, typescript_json_serializer_1.JsonProperty)({ name: "image_url3" })),
    __param(10, (0, typescript_json_serializer_1.JsonProperty)({ name: "image_url4" })),
    __param(11, (0, typescript_json_serializer_1.JsonProperty)({ name: "image_url5" })),
    __param(12, (0, typescript_json_serializer_1.JsonProperty)({ name: "app_margin_rate", required: true })),
    __param(13, (0, typescript_json_serializer_1.JsonProperty)({ name: "tenant_margin_rate", required: true })),
    __param(14, (0, typescript_json_serializer_1.JsonProperty)({ name: "min_price" })),
    __param(15, (0, typescript_json_serializer_1.JsonProperty)({ name: "max_price" })),
    __param(16, (0, typescript_json_serializer_1.JsonProperty)({ name: "price_type" })),
    __param(17, (0, typescript_json_serializer_1.JsonProperty)({ name: "payment_type", required: true })),
    __param(18, (0, typescript_json_serializer_1.JsonProperty)({ name: "tags" })),
    __param(19, (0, typescript_json_serializer_1.JsonProperty)({ name: "digital_content", required: true })),
    __metadata("design:paramtypes", [Number, Number, String, String, String, String, String, String, String, String, String, String, Number, Number, Number, Number, String, String, String, Boolean])
], BaseAppItem);
exports.BaseAppItem = BaseAppItem;
let SellerApp = class SellerApp {
    constructor(sellerAppId, baseAppId, name, shortName, description, sortOrder, iconUrl, imageUrl1, imageUrl2, imageUrl3, imageUrl4, imageUrl5, available) {
        this.sellerAppId = sellerAppId;
        this.baseAppId = baseAppId;
        this.name = name;
        this.shortName = shortName;
        this.description = description;
        this.sortOrder = sortOrder;
        this.iconUrl = iconUrl;
        this.imageUrl1 = imageUrl1;
        this.imageUrl2 = imageUrl2;
        this.imageUrl3 = imageUrl3;
        this.imageUrl4 = imageUrl4;
        this.imageUrl5 = imageUrl5;
        this.available = available;
    }
};
SellerApp = __decorate([
    (0, typescript_json_serializer_1.JsonObject)(),
    __param(0, (0, typescript_json_serializer_1.JsonProperty)({ name: "seller_app_id", required: true })),
    __param(1, (0, typescript_json_serializer_1.JsonProperty)({ name: "app_id", required: true })),
    __param(2, (0, typescript_json_serializer_1.JsonProperty)({ name: "name", required: true })),
    __param(3, (0, typescript_json_serializer_1.JsonProperty)({ name: "short_name" })),
    __param(4, (0, typescript_json_serializer_1.JsonProperty)({ name: "description" })),
    __param(5, (0, typescript_json_serializer_1.JsonProperty)({ name: "sort_order" })),
    __param(6, (0, typescript_json_serializer_1.JsonProperty)({ name: "icon_url" })),
    __param(7, (0, typescript_json_serializer_1.JsonProperty)({ name: "image_url1" })),
    __param(8, (0, typescript_json_serializer_1.JsonProperty)({ name: "image_url2" })),
    __param(9, (0, typescript_json_serializer_1.JsonProperty)({ name: "image_url3" })),
    __param(10, (0, typescript_json_serializer_1.JsonProperty)({ name: "image_url4" })),
    __param(11, (0, typescript_json_serializer_1.JsonProperty)({ name: "image_url5" })),
    __param(12, (0, typescript_json_serializer_1.JsonProperty)({ name: "available" })),
    __metadata("design:paramtypes", [Number, Number, String, String, String, Number, String, String, String, String, String, String, Boolean])
], SellerApp);
exports.SellerApp = SellerApp;
let SellerAppItem = class SellerAppItem {
    constructor(sellerAppItemId, baseAppItemId, sellerAppId, name, description, iconUrl, imageUrl1, imageUrl2, imageUrl3, imageUrl4, imageUrl5, appMarginRate, tenantMarginRate, available, sortOrder) {
        this.sellerAppItemId = sellerAppItemId;
        this.baseAppItemId = baseAppItemId;
        this.sellerAppId = sellerAppId;
        this.name = name;
        this.description = description;
        this.iconUrl = iconUrl;
        this.imageUrl1 = imageUrl1;
        this.imageUrl2 = imageUrl2;
        this.imageUrl3 = imageUrl3;
        this.imageUrl4 = imageUrl4;
        this.imageUrl5 = imageUrl5;
        this.appMarginRate = appMarginRate;
        this.tenantMarginRate = tenantMarginRate;
        this.available = available;
        this.sortOrder = sortOrder;
    }
};
SellerAppItem = __decorate([
    (0, typescript_json_serializer_1.JsonObject)(),
    __param(0, (0, typescript_json_serializer_1.JsonProperty)({ name: "seller_app_item_id", required: true })),
    __param(1, (0, typescript_json_serializer_1.JsonProperty)({ name: "app_item_id", required: true })),
    __param(2, (0, typescript_json_serializer_1.JsonProperty)({ name: "seller_app_id", required: true })),
    __param(3, (0, typescript_json_serializer_1.JsonProperty)({ name: "name", required: true })),
    __param(4, (0, typescript_json_serializer_1.JsonProperty)({ name: "description" })),
    __param(5, (0, typescript_json_serializer_1.JsonProperty)({ name: "icon_url" })),
    __param(6, (0, typescript_json_serializer_1.JsonProperty)({ name: "image_url1" })),
    __param(7, (0, typescript_json_serializer_1.JsonProperty)({ name: "image_url2" })),
    __param(8, (0, typescript_json_serializer_1.JsonProperty)({ name: "image_url3" })),
    __param(9, (0, typescript_json_serializer_1.JsonProperty)({ name: "image_url4" })),
    __param(10, (0, typescript_json_serializer_1.JsonProperty)({ name: "image_url5" })),
    __param(11, (0, typescript_json_serializer_1.JsonProperty)({ name: "app_margin_rate" })),
    __param(12, (0, typescript_json_serializer_1.JsonProperty)({ name: "tenant_margin_rate" })),
    __param(13, (0, typescript_json_serializer_1.JsonProperty)({ name: "available" })),
    __param(14, (0, typescript_json_serializer_1.JsonProperty)({ name: "sort_order", required: true })),
    __metadata("design:paramtypes", [Number, Number, Number, String, String, String, String, String, String, String, String, Number, Number, Boolean, Number])
], SellerAppItem);
exports.SellerAppItem = SellerAppItem;
let SellerAppItemId = class SellerAppItemId {
    constructor(sellerAppItemId) {
        this.sellerAppItemId = sellerAppItemId;
    }
};
SellerAppItemId = __decorate([
    (0, typescript_json_serializer_1.JsonObject)(),
    __param(0, (0, typescript_json_serializer_1.JsonProperty)({ name: "seller_app_item_id", required: true })),
    __metadata("design:paramtypes", [Number])
], SellerAppItemId);
exports.SellerAppItemId = SellerAppItemId;
let SellerAppItemPrice = class SellerAppItemPrice {
    constructor(sellerAppItemPriceId, sellerAppItemId, price, name, description, priority) {
        this.sellerAppItemPriceId = sellerAppItemPriceId;
        this.sellerAppItemId = sellerAppItemId;
        this.price = price;
        this.name = name;
        this.description = description;
        this.priority = priority;
    }
};
SellerAppItemPrice = __decorate([
    (0, typescript_json_serializer_1.JsonObject)(),
    __param(0, (0, typescript_json_serializer_1.JsonProperty)({ name: "seller_app_item_price_id", required: true })),
    __param(1, (0, typescript_json_serializer_1.JsonProperty)({ name: "seller_app_item_id", required: true })),
    __param(2, (0, typescript_json_serializer_1.JsonProperty)({ name: "price", required: true })),
    __param(3, (0, typescript_json_serializer_1.JsonProperty)({ name: "name" })),
    __param(4, (0, typescript_json_serializer_1.JsonProperty)({ name: "description" })),
    __param(5, (0, typescript_json_serializer_1.JsonProperty)({ name: "priority", required: true })),
    __metadata("design:paramtypes", [Number, Number, Number, String, String, Number])
], SellerAppItemPrice);
exports.SellerAppItemPrice = SellerAppItemPrice;
